EventListener/LogoutEventListener.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+/*
+ * This file is part of the GesdinetJWTRefreshTokenBundle package.
+ *
+ * (c) Gesdinet <http://www.gesdinet.com/>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Gesdinet\JWTRefreshTokenBundle\EventListener;
+
+use Gesdinet\JWTRefreshTokenBundle\Model\RefreshTokenManagerInterface;
+use Gesdinet\JWTRefreshTokenBundle\Request\Extractor\ExtractorInterface;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\Security\Http\Event\LogoutEvent;
+
+class LogoutEventListener
+{
+    private RefreshTokenManagerInterface $refreshTokenManager;
+    private ExtractorInterface $refreshTokenExtractor;
+    private string $tokenParameterName;
+    private array $cookieSettings;
+    private string $logout_firewall_context;
+
+    public function __construct(
+        RefreshTokenManagerInterface $refreshTokenManager,
+        ExtractorInterface $refreshTokenExtractor,
+        string $tokenParameterName,
+        array $cookieSettings,
+        string $logout_firewall_context,
+    ) {
+        $this->refreshTokenManager = $refreshTokenManager;
+        $this->refreshTokenExtractor = $refreshTokenExtractor;
+        $this->tokenParameterName = $tokenParameterName;
+        $this->cookieSettings = array_merge([
+            'enabled' => false,
+            'same_site' => 'lax',
+            'path' => '/',
+            'domain' => null,
+            'http_only' => true,
+            'secure' => true,
+            'remove_token_from_body' => true,
+        ], $cookieSettings);
+        $this->logout_firewall_context = $logout_firewall_context;
+    }
+
+    public function onLogout(LogoutEvent $event): void
+    {
+        $request = $event->getRequest();
+        $current_firewall_context = $request->attributes->get('_firewall_context');
+
+        if ($current_firewall_context !== $this->logout_firewall_context) {
+            return;
+        }
+
+        $tokenString = $this->refreshTokenExtractor->getRefreshToken($request, $this->tokenParameterName);
+        if (null === $tokenString) {
+            $event->setResponse(
+                new JsonResponse(
+                    [
+                        'code' => 422,
+                        'message' => 'No refresh_token found.',
+                    ],
+                    JsonResponse::HTTP_UNPROCESSABLE_ENTITY
+                )
+            );
+        } else {
+            $refreshToken = $this->refreshTokenManager->get($tokenString);
+            if (null === $refreshToken) {
+                $event->setResponse(
+                    new JsonResponse(
+                        [
+                            'code' => 422,
+                            'message' => 'Invalid refresh_token supplied.',
+                        ],
+                        JsonResponse::HTTP_UNPROCESSABLE_ENTITY
+                    )
+                );
```

Comment:
Now that I think about it more, I wonder if a 4xx response is even appropriate in this case.  If you're trying to invalidate a token that doesn't exist in the system, the desired end state (invalidated token) is still reached.  It's not much different than ignoring errors if the file you're trying to delete from the filesystem already doesn't exist.

```php
try {
    $this->filesystem->delete($file);
} catch (FileNotFoundException) {
    // If it's gone then we're OK
}
```

Reply:
Makes sense. I changed this to send a 200 response indicating the supplied refresh token is already invalid. It still falls through and unsets the cookie.